### ISSUE TITLE
Total files size is limited by system memory: memory usage in increased until all system memory is used and the program is killed

### ISSUE THREAD

**Issue by AIGeneratedUsername** [Lymphatus/caesium-image-compressor#178]
**Describe the bug**

When running against a large folder, memory usage in increased until all system memory is used and the program is killed.

I can imagine one of two possible problems:

1. There is a memory leak.
2. The program uses memory inefficiently by putting all files at once into memory instead of processing files in batches. Pseudo-code:
    ```python
    # Do not put all files into memory by processing no more then N files in parallel
    for chunk in chunked(files, chunk_size=n):
        for file in chunk:
            process(file)
    ```

**Software version**
Caesium Image Compressor 2.2.1

**Operating System information**
 - Lubuntu 20.04

**To Reproduce**
1. Run against a 30 GB folder with a system which has 8 GB memory + 16 GB swap file.

Wait about for 1 hour until all memory will be filled and the program will be killed.

**Expected behavior**

Memory-efficient behavior.

**Application Log**
The log has only "Preview in..." info messages which are probably not useful now. I can attach it as a file, but do I really need to attach a full log?
```
[2023-01-24 10:18:00.411][I] Preview in "/home/sam/.cache/SaeraSoft/Caesium Image Compressor/8f3886b0e92ecc0e588c556716a84c72dbd65036eb9b7577652ea179ec38f177"
[2023-01-24 10:18:00.604][I] Preview in "/home/sam/.cache/SaeraSoft/Caesium Image Compressor/e6b430b2a27be673ad253a13af308149ea6e66d4dfcfbcfc7949b4881fe5a282"
[2023-01-24 10:35:36.656][I] ---- Starting application ----
[2023-01-24 10:35:36.894][I] Starting UI
```

### Workarounds

I just found that there is a CLI: https://github.com/Lymphatus/caesium-clt

Which means that I can write a Python (or bash) script which will call Caesium for every nested directory one by one.

**Comment by Lymphatus** [comment]
It might be a problem with the Linux OS. It's not as tested as the Windows and macOS releases and might have some kind of memory problem. I have reports of compressions for over 1Tb of data all at once, so 30Gb should be a breeze.
I'll take a look into that.

By the way, the caesium-clt utility can compress folders and subfolders too using the `-RS` option. If you have any doubts or want to know more, we can discuss in its repo.

**Comment by AIGeneratedUsername** [comment]
Thank you.

One more possible problem I found is that memory is not released when processing is done and the program is just open without any action.

Here is an open program which does not process any files:
![memory_usage](https://user-images.githubusercontent.com/70140627/214387722-24e8cf6d-7a27-41f1-8b61-736a24d6648a.png)

Right after closing:
![memory_after](https://user-images.githubusercontent.com/70140627/214387773-e48e5602-611a-4bdf-ad6f-eafa08739a05.png)

**Comment by xukgo** [comment]
This problem also occurred when I integrated the rust libcaesium into my project and traversed a folder which has many image files. Through the exclusion method, I commented the compress lib function that after the compression operation, the memory did not exceed 5M after traversing, so it should be the rust library problem. Caused by not releasing some resources.

**Comment by Lymphatus** [comment]
@xukgo yes it is. I investigated further in the libcaesium code and found out there was a major memory leak in the JPG compress function. I released a fix to that in https://github.com/Lymphatus/libcaesium/commit/2b71a83c0b365c4377717feb405658b60d1dec48.
It should be shipped within the next Caesium release.

**Comment by Lymphatus** [comment]
Fixed in 2.4.0
